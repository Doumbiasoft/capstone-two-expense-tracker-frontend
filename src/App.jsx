import React,{useState, useEffect} from 'react';
import { CssBaseline, ThemeProvider } from '@mui/material';
import { useSelector } from 'react-redux';
import RTL from './layouts/full-layout/customizer/RTL';
import ThemeSettings from './layouts/full-layout/customizer/ThemeSettings';
import RouterManager from './routes/Router';
import 'react-perfect-scrollbar/dist/css/styles.css';
import "slick-carousel/slick/slick.css";
import "slick-carousel/slick/slick-theme.css";
import Api from './api';
import { useLocalStorage } from './hooks/useLocalStorage'; 
import CurrentUserContext from './contexts/CurrentUserContext';
import './App.scss';

function App() {
  const theme = ThemeSettings();
  const customizer = useSelector((state) => state.CustomizerReducer);

  const [user, setUser] = useState(null);
  const [userId, seUserId] = useState(0);
  const [token, setToken] = useLocalStorage("token","");

  const handleUser = (data) =>{
    setUser(data);
  }
  const handleToken = (_token) =>{
    setToken(_token);
  }

  const handleLogOut = () =>{
    setToken("");
    setUser(null);
  }

  useEffect(() =>{
    Api.token = token;
  },[user,token]);

  useEffect(() => {
    /* get the current */
    async function getCurrentUser(){
      try {
        if (user === null) {
          if (token){
            const tokenInfo = Api.decodeToken(token);
            seUserId(tokenInfo.id);
            if(tokenInfo){
                const currentUser = await Api.getUser(tokenInfo.id);
                if(currentUser){
                  setUser(currentUser);
                }else{
                  handleLogOut();
                }
            }
          } 
        }
      } catch (error) {
        handleLogOut();
      }
    }
    getCurrentUser();
   }, [user, token]);


  const ctx = {
    userId:userId,
    user:{...user}, 
    token:token, 
    actions:{ 
        handleUser:handleUser,
        handleToken:handleToken,
        handleLogOut:handleLogOut, 
      }
   
    };
  return (
    <CurrentUserContext.Provider value={ctx}>
    <ThemeProvider theme={theme}>
      <RTL direction={customizer.activeDir}>
        <CssBaseline />
        {<RouterManager />}
      </RTL>
    </ThemeProvider>
    </CurrentUserContext.Provider>
  );
}

export default App
