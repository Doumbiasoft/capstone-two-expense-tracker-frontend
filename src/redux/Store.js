import { createStore } from 'redux';
import RootReducers from './Rootreducers';

export function configureStore(InitialState) {
  const Store = createStore(
    RootReducers,
    InitialState,
  );
  return Store;
}
