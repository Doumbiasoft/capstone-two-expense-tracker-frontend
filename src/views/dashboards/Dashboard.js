import React,{useEffect,useState,useContext} from 'react';
import { Grid } from '@mui/material';
import {
  AmountTile,
  DoughnutChartExpenseByCategory,
  LineChartExpenseVsIncome,
  RecentTransactions
} from './dashboard-components';
import Breadcrumb from '../../layouts/full-layout/breadcrumb/Breadcrumb';
const BCrumb = [
  {
    title: 'Home',
  },
  {
    title: 'Dashboard',
  }
];

import PageContainer from '../../components/container/PageContainer';
import Api from '../../api'
import CurrentUserContext from '../../contexts/CurrentUserContext';


export default function Dashboard() {

  const ctx = useContext(CurrentUserContext);
  const [dashboard, setDashboard] = useState({});

  async function getDashboard() {
    try {
       const data = await Api.getDashboard(ctx.userId);
        if (data){
          setDashboard(data);
        }
      } catch (error) {
    }
}
useEffect(() => {
  getDashboard();
},[]);

  return (
    
  <PageContainer title="Dashboard" description="this is Analytical Dashboard">
  {/* breadcrumb */}
  <Breadcrumb title="Dashboard" items={BCrumb} />
  <Grid container spacing={0}>
    {/* ------------------------- row 1 ------------------------- */}

    <Grid item xs={12} sm={6} lg={4}>
      <AmountTile title='Total Income' amount={dashboard.totalIncome?dashboard.totalIncome:0} fabColor='success' titleColor='#00C292' />
    </Grid>
    <Grid item xs={12} sm={6} lg={4}>
    <AmountTile title='Total Expense' amount={dashboard.totalExpense?dashboard.totalExpense:0} fabColor='danger' titleColor='#E46A76'/>
    </Grid>
    <Grid item xs={12} sm={12} lg={4}>
    <AmountTile title='Total Balance' amount={dashboard.balance?dashboard.balance:0} fabColor='primary' titleColor='#364AAE'/>
    </Grid>
    <Grid item xs={12} lg={8}>
      <LineChartExpenseVsIncome splineChartData={dashboard.splineChartData?dashboard.splineChartData:[]} />
    </Grid>
    <Grid item xs={12} lg={4}>
      <DoughnutChartExpenseByCategory doughnutChartData={dashboard.doughnutChartData?dashboard.doughnutChartData:[]} />
    </Grid>
  
    <Grid item xs={12} lg={12}>
      <RecentTransactions data={dashboard.recentTransactions?dashboard.recentTransactions:[]} />
    </Grid>
    
  </Grid>
</PageContainer>
  )
}
