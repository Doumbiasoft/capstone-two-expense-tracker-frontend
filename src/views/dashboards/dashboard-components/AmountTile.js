import React from 'react';
import { Card, CardContent, Typography, Box, Fab } from '@mui/material';

import FeatherIcon from 'feather-icons-react';
import cultureInfo from '../../../helpers/cultureInfo';


export default function AmountTile({title, amount, fabColor="secondary", titleColor}) {
  return (
    <Card>
    <CardContent>
      <Box display="flex" alignItems="flex-start">
        <Typography
          variant="h4"
          sx={{
            marginBottom: '0',
            
          }}
          gutterBottom
          color={titleColor}
        >
          {title}
        </Typography>
        <Box
          sx={{
            marginLeft: 'auto',
          }}
        >
          <Fab
            size="medium"
            color={fabColor}
            aria-label="add"
            elevation="0"
            sx={{
              boxShadow: 'none',
            }}
          >
            <FeatherIcon icon="dollar-sign" size={30} />
          </Fab>
        </Box>
      </Box>
      <Typography
        variant="h1"
        fontWeight="600"
        sx={{
          marginBottom: '0',
          marginTop: '20px',
        }}
        gutterBottom
      >
        {cultureInfo.format(amount)}
      </Typography>
     
    </CardContent>
  </Card>
  )
}

