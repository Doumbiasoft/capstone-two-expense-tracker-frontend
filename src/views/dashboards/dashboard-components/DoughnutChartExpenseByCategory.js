
import React from 'react';
import { Box, Typography } from '@mui/material';
import Chart from 'react-apexcharts';
import DashboardCard from '../../../components/base-card/DashboardCard';

const DoughnutChartExpenseByCategory = ({doughnutChartData}) => {

  const labels = doughnutChartData.map(data => data.categoryName);
  const seriesChart = doughnutChartData.map(data => data.amount);

  const COLORS = ['#00C49F','#364AAE','#F2BE4E','#EF9264','#4E5AA0','#97BD3F','#A8108F', '#0088FE', '#CD8DF5'];

  const options = {
    labels: labels,

    chart: {
      height: 220,
      type: 'donut',
      foreColor: '#adb0bb',
      fontFamily: 'DM sans',
    },
    colors: COLORS,
    dataLabels: {
      enabled: true,
    },
    legend: {
      show: true,
    },
    stroke: {
      colors: ['transparent'],
    },
    plotOptions: {
      pie: {
        donut: {
          size: '50%',
          background: 'transparent',
          labels: {
            show: true,
            name: {
              show: true,
              fontSize: '18px',
              color: undefined,
              offsetY: -10,
            },
            value: {
              show: true,
              color: '#98aab4',
            },
            total: {
              show: false,
              label: 'Total',
              color: '#98aab4',
            },
          },
        },
      },
    },
    tooltip: {
      theme: 'dark',
      fillSeriesColor: true,
    },
  };
  const series = seriesChart;
  return (
    <DashboardCard title="Expense By Category">
 <Typography
          color="textSecondary"
          variant="body1"
          sx={{
            fontSize: 'h5.fontSize',
          }}
        >
          Last 7 days
        </Typography>
      <Box
        sx={{
          mt: 5,
          position: 'relative',
          height: 220,
        }}
      >
        <Chart options={options} series={series} type="donut" height="220" width="100%" />
 
      </Box>
      
    </DashboardCard>
  );
};

export default DoughnutChartExpenseByCategory;
