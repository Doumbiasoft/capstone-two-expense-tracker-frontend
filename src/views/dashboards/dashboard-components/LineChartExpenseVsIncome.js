import React from 'react';
import { Box, Typography } from '@mui/material';
import Chart from 'react-apexcharts';
import DashboardCard from '../../../components/base-card/DashboardCard';

export default function LineChartExpenseVsIncome({ splineChartData }) {


  const categories = (Array.isArray(splineChartData) && splineChartData.length > 0) ? splineChartData.map(data => data.day) : [];
  const incomeData = (Array.isArray(splineChartData) && splineChartData.length > 0) ? splineChartData.map(data => data.income) : [];
  const expenseData = (Array.isArray(splineChartData) && splineChartData.length > 0) ? splineChartData.map(data => data.expense) : [];


  const primary = "#7CCF71";
  const secondary = "#E7D91B";
  const options = {
    grid: {
      show: true,
      borderColor: 'rgba(0, 0, 0, .2)',
      color: '#777e89',
      strokeDashArray: 2,
      xaxis: {
        lines: {
          show: false,
        },
      },
      yaxis: {
        lines: {
          show: true,
        },
      },
    },
    dataLabels: {
      enabled: false,
    },
    chart: {
      fontFamily: 'DM Sans',
      foreColor: '#adb0bb',
      toolbar: {
        show: false,
      },
    },
    stroke: {
      curve: 'smooth',
      width: 3,
    },
    colors: [primary, secondary],

    xaxis: {
      categories: categories,
    },
    markers: {
      size: 4,
      border: 1,
    },
    tooltip: {
      x: {
        format: 'dd/MM/yy HH:mm',
      },
      theme: 'dark',

    },
    legend: {
      show: true,
    },
  };
  const series = [
    {
      name: 'Income',
      data: incomeData,
    },
    {
      name: 'Expense',
      data: expenseData,
    },
  ];
  return (
    <DashboardCard title="Income vs Expense">
      {/* chart */}
      <Typography
        color="textSecondary"
        variant="body1"
        sx={{
          fontSize: 'h5.fontSize',
        }}
      >
        Last 7 days
      </Typography>
      <Box>
        <Chart options={options} series={series} type="line" height="245" />
      </Box>
    </DashboardCard>
  );
}
