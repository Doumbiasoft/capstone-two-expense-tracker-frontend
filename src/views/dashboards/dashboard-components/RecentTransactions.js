import React from 'react';
import {
  Typography,
  Box,
  Table,
  TableBody,
  TableCell,
  TableHead,
  TableRow,
  Chip,
} from '@mui/material';
import DashboardCard from '../../../components/base-card/DashboardCard';
import cultureInfo from '../../../helpers/cultureInfo';
import displayDateFormat from '../../../helpers/displayDateFormat';
//const data =[{id:"1",category:"Salary",date:"14 Dec 2019",type:"Income",amount:"21000"},{id:"2",category:"Transportation",date:"04 Dec 2019",type:"Expense",amount:"500"}]


export default function RecentTransactions({data}) {


  return (
<DashboardCard
    title="Last 5 Recent Transactions"
    subtitle=""
    customdisplay="block"
    custommargin="10px"
  >
    <Box
      sx={{
        overflow: 'auto',
        mt: -3,
      }}
    >
      <Table
        aria-label="simple table"
        sx={{
          mt: 2,
          whiteSpace: 'nowrap',
          'td:first-of-type, th:first-of-type': {
            pl: 0,
          },
          'td:last-of-type, th:last-of-type': {
            pr: 0,
          },
        }}
      >
        <TableHead>
          <TableRow>
            <TableCell>
              <Typography variant="h5" color="textSecondary">
                Category
              </Typography>
            </TableCell>
            <TableCell>
              <Typography variant="h5" color="textSecondary">
                Date
              </Typography>
            </TableCell>
            <TableCell>
              <Typography variant="h5" color="textSecondary">
                Amount
              </Typography>
            </TableCell>
            <TableCell align="right">
              <Typography variant="h5" color="textSecondary">
                Type
              </Typography>
            </TableCell>
          </TableRow>
        </TableHead>
        <TableBody>
          {data.map((row) => (
            <TableRow key={row.id}>
               <TableCell>
                <Typography color="textSecondary" variant="h6">
                  {row.categoryName}
                </Typography>
              </TableCell>
              <TableCell>
                <Typography color="textSecondary" variant="h6">
                  {row.date}
                </Typography>
              </TableCell>
              <TableCell>
                <Typography color="textSecondary" variant="h6">
                  <span style={{color:row.categoryType === "Expense"?"red":"green"}}>{row.categoryType === 'Expense'?'- ':"+ "}</span>{cultureInfo.format(row.amount)}
                </Typography>
              </TableCell>
              <TableCell>
              <Box display="flex" justifyContent='end'>
                <Chip
                  sx={{
                    backgroundColor:
                    row.categoryType === 'Expense'
                        ? (theme) => theme.palette.warning.main
                        : (theme) => theme.palette.success.main,
                    color: '#fff',
                    borderRadius: '6px',
                
                  }}
                  size="small"
                  label={row.categoryType}
                />
                </Box>
              </TableCell>
        
            </TableRow>
          ))}
        </TableBody>
      </Table>
    </Box>
  </DashboardCard>
  )
}

