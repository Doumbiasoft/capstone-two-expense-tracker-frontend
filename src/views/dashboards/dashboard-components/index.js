
import AmountTile from './AmountTile';
import DoughnutChartExpenseByCategory from './DoughnutChartExpenseByCategory';
import LineChartExpenseVsIncome from './LineChartExpenseVsIncome';
import RecentTransactions from './RecentTransactions';

export {
  AmountTile,
  DoughnutChartExpenseByCategory,
  LineChartExpenseVsIncome,
  RecentTransactions
};
