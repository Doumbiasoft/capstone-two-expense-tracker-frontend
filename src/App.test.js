import { render, screen } from '@testing-library/react';
import { describe, it, expect, vi } from 'vitest';
import { Provider } from 'react-redux';
import { configureStore } from './redux/Store';
import { BrowserRouter,MemoryRouter } from 'react-router-dom';
import App from './App';
import { AmountTile } from './views/dashboards/dashboard-components';
import { RecentTransactions } from './views/dashboards/dashboard-components';

describe("Smoke Test App", () => {
  

  it('Should not crash when rendering App', async () => {
    const mockStore = configureStore();
    let store;
    store = mockStore;
    render(
      <Provider store={store}>
        <BrowserRouter>
          <App />
        </BrowserRouter>
      </Provider>
    );
  });

});


describe("Smoke Test Dashboard", () => {

  it('should not crash when rendering AmountTile', async () => {
    const amount = vi.fn();
    const fabColor = vi.fn();
    const titleColor = vi.fn();
    render(<AmountTile title='Total Income' amount={amount} fabColor={fabColor} titleColor={titleColor} />);
  });

  it('should not crash when rendering Recent Transaction', async () => {
    const data = vi.fn();
    render(<RecentTransactions data={data} />);
  });
  

});

describe("Snapshot Test", async () => {

  it("should matches snapshot of AmountTile", () => {
    const amount = vi.fn();
    const fabColor = vi.fn();
    const titleColor = vi.fn();
    const { asFragment } = render(
      <AmountTile title='Total Income' amount={amount} fabColor={fabColor} titleColor={titleColor} />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should matches snapshot Recent Transactions", () => {
    const data = vi.fn();
    const { asFragment } = render(<RecentTransactions data={data} />);
    expect(asFragment()).toMatchSnapshot();
  });


});