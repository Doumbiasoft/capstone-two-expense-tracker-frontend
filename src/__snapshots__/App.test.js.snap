// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Snapshot Test > should matches snapshot Recent Transactions 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-qxkz57-MuiPaper-root-MuiCard-root"
  >
    <div
      class="MuiCardHeader-root css-v8ux6d-MuiCardHeader-root"
      subtitle=""
    >
      <div
        class="MuiCardHeader-content css-1qbkelo-MuiCardHeader-content"
      >
        <h3
          class="MuiTypography-root MuiTypography-h3 css-vqiuom-MuiTypography-root"
        >
          Recent Transactions
        </h3>
      </div>
    </div>
    <div
      class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 css-he9jqb-MuiTypography-root"
      >
        Last 5
      </p>
      <div
        class="MuiBox-root css-1ap4dyu"
      >
        <table
          aria-label="simple table"
          class="MuiTable-root css-lmmtxd-MuiTable-root"
        >
          <thead
            class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
          >
            <tr
              class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
            >
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
                scope="col"
              >
                <h5
                  class="MuiTypography-root MuiTypography-h5 css-41an0y-MuiTypography-root"
                >
                  Category
                </h5>
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
                scope="col"
              >
                <h5
                  class="MuiTypography-root MuiTypography-h5 css-41an0y-MuiTypography-root"
                >
                  Date
                </h5>
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
                scope="col"
              >
                <h5
                  class="MuiTypography-root MuiTypography-h5 css-41an0y-MuiTypography-root"
                >
                  Amount
                </h5>
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-1azl6jz-MuiTableCell-root"
                scope="col"
              >
                <h5
                  class="MuiTypography-root MuiTypography-h5 css-41an0y-MuiTypography-root"
                >
                  Type
                </h5>
              </th>
            </tr>
          </thead>
          <tbody
            class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
          />
        </table>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Snapshot Test > should matches snapshot of AmountTile 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
  >
    <div
      class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
    >
      <div
        class="MuiBox-root css-1yp4ln"
      >
        <h4
          class="MuiTypography-root MuiTypography-h4 MuiTypography-gutterBottom css-1odcp3x-MuiTypography-root"
        >
          Total Income
        </h4>
        <div
          class="MuiBox-root css-zdpt2t"
        >
          <button
            aria-label="add"
            class="MuiButtonBase-root MuiFab-root MuiFab-circular MuiFab-sizeMedium MuiFab-function(...a) {
    let r = v(t);
    r.called = !0, r.callCount++, r.calls.push(a);
    let i = r.next.shift();
    if (i) {
      r.results.push(i);
      let [s, l] = i;
      if (s === \\"ok\\")
        return l;
      throw l;
    }
    let o, c = \\"ok\\";
    if (r.impl)
      try {
        new.target ? o = Reflect.construct(r.impl, a, new.target) : o = r.impl.apply(this, a), c = \\"ok\\";
      } catch (s) {
        throw o = s, c = \\"error\\", r.results.push([c, s]), s;
      }
    let x = [c, o];
    if (b(o)) {
      let s = o.then((l) => x[1] = l).catch((l) => {
        throw x[0] = \\"error\\", x[1] = l, l;
      });
      Object.assign(s, o), o = s;
    }
    return r.results.push(x), o;
  } MuiFab-root MuiFab-circular MuiFab-sizeMedium MuiFab-function(...a) {
    let r = v(t);
    r.called = !0, r.callCount++, r.calls.push(a);
    let i = r.next.shift();
    if (i) {
      r.results.push(i);
      let [s, l] = i;
      if (s === \\"ok\\")
        return l;
      throw l;
    }
    let o, c = \\"ok\\";
    if (r.impl)
      try {
        new.target ? o = Reflect.construct(r.impl, a, new.target) : o = r.impl.apply(this, a), c = \\"ok\\";
      } catch (s) {
        throw o = s, c = \\"error\\", r.results.push([c, s]), s;
      }
    let x = [c, o];
    if (b(o)) {
      let s = o.then((l) => x[1] = l).catch((l) => {
        throw x[0] = \\"error\\", x[1] = l, l;
      });
      Object.assign(s, o), o = s;
    }
    return r.results.push(x), o;
  } css-1ts8yrn-MuiButtonBase-root-MuiFab-root"
            elevation="0"
            tabindex="0"
            type="button"
          >
            <svg
              class="feather feather-dollar-sign "
              fill="none"
              height="30"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="30"
            >
              <g>
                <line
                  x1="12"
                  x2="12"
                  y1="1"
                  y2="23"
                />
                <path
                  d="M17 5H9.5a3.5 3.5 0 0 0 0 7h5a3.5 3.5 0 0 1 0 7H6"
                />
              </g>
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
      <h1
        class="MuiTypography-root MuiTypography-h1 MuiTypography-gutterBottom css-69kgue-MuiTypography-root"
      >
        $NaN
      </h1>
      <h6
        class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom css-1c9hwm4-MuiTypography-root"
      >
        Monthly Income
      </h6>
    </div>
  </div>
</DocumentFragment>
`;
